staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,595 @@
+		// technique to pick a queue, dequeue the request at the head of that
+		// queue, increment the count of the number executing, and send true to
+		// the request's channel.
+		qs.dequeueWithChannelAsMuchAsPossible()
+		// ------------------------------------------------------------------------
+		return false, false, false, func() {}
+	}()
+	if shouldReturn {
+		return tryAnother, execute, afterExecution
+	}
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.name, "time-out")
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request cancelled\n")
+		func() {
+			qs.lockAndSyncTime()
+			defer qs.lock.Unlock()
+
+			// TODO(aaron-prindle) add metrics to these two cases
+			if req.Enqueued {
+				// remove the request from the queue as it has timed out
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						// remove the request
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+					qs.getRequestsExecutingLocked() == 0 {
+					qs.counter.Add(1)
+					go func(eh fq.EmptyHandler) {
+						eh.HandleEmpty()
+						qs.counter.Add(-1)
+					}(qs.emptyHandler)
+				}
+			} else {
+				// At this point we know that req was in its queue earlier and another
+				// goroutine has removed req from its queue and called qs.counter.Add(1)
+				// in anticipation of unblocking this goroutine through the other arm of this
+				// select. In this case we need to decrement the counter because this goroutine
+				// was actually unblocked through a different code path.
+				qs.counter.Add(-1)
+			}
+		}()
+		return false, false, func() {}
+	}
+	// ************************************************************************
+}
+
+// syncTimeLocked is used to sync the time of the queueSet by looking at the elapsed
+// time since the last sync and this value based on the 'virtualtime ratio'
+// which scales inversely to the # of active flows
+func (qs *queueSet) syncTimeLocked() {
+	realNow := qs.clock.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	var virtualTimeRatio float64
+
+	emptyCount := 0
```

Comment:
This variable should not be named "emptyCount", it is not a count of empty anythings.  The logic shows that it is a count of the number of _non_ empty (in the broad sense that includes executing as well as waiting requests) queues.

Reply:
Updated - changed from s/emptyCount/activeQueues.